File: Arduino/Libraries/RACOM/RACOM_DL.cpp
#include "RACOM_DL.h"

RACOM_DL::RACOM_DL() { 
}

void RACOM_DL::begin() {
  _available = 0;
  _pSize = 0;
  _cnt = 0;
  _packet = malloc(MAX_FDATA_SIZE+2);
  RacomPHY.begin();
  
}

void RACOM_DL::end() {
  free(_packet);
  _packet = NULL;
  RacomPHY.end();
}

uint8_t RACOM_DL::send(uint8_t pSize, uint8_t* packet) {
  RacomPHY.write(_STX);
  RacomPHY.write((uint8_t)(pSize<<2) + (packet[0] & 0b11));
  if(pSize==2)
    RacomPHY.write(*(packet+1));
  else if(pSize>2)
    RacomPHY.write((packet+1),(pSize-1));
}

int8_t RACOM_DL::available(){
  this->readSM();
  return _available;
}

uint8_t RACOM_DL::pSize(){
  if(_available==1)
    return _pSize;
  else
    return 0;
}

int8_t RACOM_DL::read(uint8_t* pSize, uint8_t* packet) {
  if(_available==1){
    *pSize = _pSize;
    for(uint8_t i = 0;i<_pSize;i++){
      packet[i]=_packet[i];
    }
    _available = 0;
    return 1;
  }
  else
    return _available;
}

void RACOM_DL::flushRx() {
  while(RacomPHY.available()!=0)
    RacomPHY.read();
  this->_available = 0;
}

int8_t RACOM_DL::readSM() {
  static uint8_t _state = _WAITING_STATE;
  static uint32_t _t0;
  
  switch(_state) {
    case _WAITING_STATE: {
      uint8_t tmp = 0;
      while(RacomPHY.available()!=0){
        tmp = RacomPHY.read();
        if(tmp==_STX) {
          _cnt = 0;
          break;
        }
      }
      if(tmp==_STX)
        _state = _START_STATE;
    }
    break;
      
    case _START_STATE:
      _t0 = millis();
      if(RacomPHY.available()>0 && _available <= 0) {
        uint8_t tmp = RacomPHY.read();
        _pSize = tmp >> 2;
        _packet[0] = tmp & 0b11;
        _cnt = 1;
      }
      if(_cnt == 1) {
        _state = _WAIT_DATA_STATE;
      }
      else if((millis()-_t0)>= TIMEOUT){
        _state = _WAITING_STATE;
        _available = -1;
        return -1;
      }
      break;
      
    case _WAIT_DATA_STATE:
      while(RacomPHY.available()>0&&_cnt<_pSize) {
        _packet[_cnt] = RacomPHY.read();
        _cnt+=1;
      }
      if((millis()-_t0)>= TIMEOUT){
        _state = _WAITING_STATE;
        _available = -1;
        return -1;
      }
      else if(_cnt==(_pSize)){
        _available = 1;
        _state = _WAITING_STATE;
        return 1;
      }
      break;      
  }
  return 0;
}

#ifdef UART_INTERFACE
void serialEvent(){
  if(RacomPHY.available()>0){
    #ifdef DL_EN
      RacomDL.readSM();
    #endif
  }
}
#elif defined(I2C_INTERFACE)

#endif

RACOM_DL RacomDL = RACOM_DL();

File: ROS_Casper/src/Libraries/RACOM/RACOM_DL.py
#!/usr/bin/env python

import logging
import RACOM_PHY
from datetime import datetime
from time import sleep


class RACOM_DL(object):
    # Com properties
    _TIMEOUT = 0.5
    #Frame properties
    _STX = 0x02  # start

    #SM variables
    _t0 = 0
    _state = 0
    _packet = []
    _pSize = 0
    _cnt = 0
    _available = 0

    MAX_PDATA_SIZE = 60
    
    def __init__(self,interfaceName):
        """
        Constructor of the Framing class
        """
        self.baudrate = 115200
        self.logger = logging.getLogger(__name__)
        if interfaceName == "UART":
            self.RacomPHY = RACOM_PHY.UART(self.baudrate,self._TIMEOUT)
        elif interfaceName == "I2C":
            self.RacomPHY = RACOM_PHY.I2C(8)
        
    def formatPacket(self, packet):
        
        frame = []
        frame.append(self._STX)
        frame.append((len(packet)<<2) + (packet[0] & 0b11))

        ##DATA
        for i in range(1, len(packet), 1):
            if type(packet[i]) is int:
                frame.append(packet[i])
            else:
                frame.append(ord(packet[i]))
        return frame

    def send(self,packet):
        frame = self.formatPacket(packet)
        self.RacomPHY.write(frame)
		sleep(0.5)

    def read(self):
        if self._available == 1:
            self._available = 0
            return self._packet
        else:
            return 0

    def pSize(self):
        return self._pSize
        
    def available(self):
        if(self.readSM()==-1):
            return -1
        return self._available
    
    def readSM(self):
        #States
        _WAITING_STATE = 0
        _START_STATE = 1
        _WAIT_DATA_STATE = 2

        if self._state is _WAITING_STATE:
            byte = []
            while self.RacomPHY.available() != 0 :
                byte = self.RacomPHY.read()
                if len(byte) == 0:
                    return 0
                else:
                    byte=byte[0]
                if byte is self._STX:
                    self._cnt = 0
                    break
                if self.RacomPHY.available() == -1 :
					break
            if byte is self._STX :
                    self._state = _START_STATE

        elif self._state is _START_STATE:
            self._t0 = datetime.now()
            if self.RacomPHY.available() > 0 and self._available == 0:
                self._packet = []
                self._pSize = self.RacomPHY.read()[0]
                self._pFlags = self._pSize & 0b11
                self._pSize = self._pSize >> 2
                self._packet.append(self._pFlags)
                self._cnt = 1
            elif self.RacomPHY.available() != 0 and self._available == 0:
                self._packet = []
                try:
                    self._pSize = self.RacomPHY.read()[0]
                except:
                    return 0
                self._pFlags = self._pSize & 0b11
                self._pSize = self._pSize >> 2
                self._packet.append(self._pFlags)
                self._cnt = 1
            if self._cnt == 1:
                self._state = _WAIT_DATA_STATE
            elif (datetime.now()-self._t0).total_seconds() >= self._TIMEOUT :
                self._state = _WAITING_STATE
                return -1   
                

        elif self._state is _WAIT_DATA_STATE:
            while self.RacomPHY.available() > 0 and self._cnt < self._pSize:
                self._packet[len(self._packet):]=self.RacomPHY.read()
                self._cnt+=1
            if self.RacomPHY.available() == -1 and self._cnt < self._pSize:
                self._packet[len(self._packet):]=self.RacomPHY.readBytes(self._pSize-1)
                self._cnt+=self._pSize-1
            if (datetime.now()-self._t0).total_seconds() >= self._TIMEOUT :
                self._state = _WAITING_STATE
                return -1
            elif self._cnt == self._pSize :
                self._available = 1
                self._state = _WAITING_STATE
                return 1
        
        return 0